File: matlab/examples/analyseSpeech.m
clear
clc
close all

% add the estimator object to the MATLAB path
addpath ../fastF0Nls

% load the mono speech signal
[speechSignal, samplingFreq] = audioread('roy.wav');
nData = length(speechSignal);

% set up
segmentTime = 0.025; % seconds
segmentLength = round(segmentTime*samplingFreq); % samples
nSegments = floor(nData/segmentLength);
f0Bounds = [80, 400]/samplingFreq; % cycles/sample
maxNoHarmonics = 10;
f0Estimator = fastF0Nls(segmentLength, maxNoHarmonics, f0Bounds);

% do the analysis
idx = 1:segmentLength;
f0Estimates = nan(1,nSegments); % cycles/sample
for ii = 1:nSegments
    disp(['Processing segment ', num2str(ii), ' of ', ...
        num2str(nSegments)]);
    speechSegment = speechSignal(idx);
    f0Estimates(ii) = f0Estimator.estimate(speechSegment);
    idx = idx + segmentLength;
end
timeVector = (1:nSegments)*segmentTime-segmentTime/2;

%% compute the spectrogram of the signal
window = gausswin(segmentLength);
nOverlap = round(3*segmentLength/4);
nDft = 2048;
[stft, stftFreqVector, stftTimeVector] = ...
    spectrogram(speechSignal, window, nOverlap, nDft, samplingFreq);
powerSpectrum = abs(stft).^2;

%% plot the results
maxDynamicRange = 60; % dB
imagesc(stftTimeVector, stftFreqVector, ...
    10*log10(dynamicRangeLimiting(powerSpectrum, maxDynamicRange)));
set(gca,'YDir','normal')
hold on
plot(timeVector, f0Estimates*samplingFreq, 'r.')
hold off
colorbar
xlabel('time [s]')
ylabel('frequency [Hz]')
title('Why where you away a year, Roy?')

File: python/example/analyseSpeech.py
import sys
from scipy.io import wavfile
import numpy as np
import scipy as sp
import scipy.signal as signal
import matplotlib.pyplot as plt

sys.path.append('../')
import single_pitch

[samplingFreq, speechSignal] = wavfile.read('roy.wav')
nData = speechSignal.shape[0]

# set up
segmentTime = 0.025 # seconds
segmentLength = round(segmentTime*samplingFreq) # samples
nSegments = int(np.floor(nData/segmentLength))
f0Bounds = np.array([80, 400])/samplingFreq
maxNoHarmonics = 15
f0Estimator = single_pitch.single_pitch(segmentLength, maxNoHarmonics, f0Bounds)

# do the analysis
idx = np.array(range(0, segmentLength))
f0Estimates = np.zeros((nSegments,)) # cycles/sample
for ii in range(nSegments):
    speechSegment = np.array(speechSignal[idx], dtype=np.float64)
    f0Estimates[ii] = (samplingFreq/(2*np.pi))*f0Estimator.est(speechSegment)
    idx = idx + segmentLength

timeVector = np.array(range(1, nSegments+1))*segmentTime-segmentTime/2

# compute the spectrogram of the signal
nOverlap = round(3*segmentLength/4)
[stftFreqVector, stftTimeVector, stft] = signal.spectrogram(speechSignal,
                                                            fs=samplingFreq,
                                                            nperseg=segmentLength,
                                                            noverlap=nOverlap, nfft=2048)
powerSpectrum = np.abs(stft)**2;

# plot the results
maxDynamicRange = 60 # dB
plt.pcolormesh(stftTimeVector, stftFreqVector, 10*np.log10(powerSpectrum))
plt.scatter(timeVector, f0Estimates, c='b', s=20)
plt.title('Why where you away a year, Roy?')
plt.xlabel('time [s]')
plt.ylabel('frequency [Hz]')
plt.show()
